extensions/multiline.md

```diff
@@ -0,0 +1,241 @@
+
+* `MULTILINE_MAX_BYTES <limit>`: `max-bytes` limit exceeded
+* `MULTILINE_MAX_LINES <limit>`: `max-lines` limit exceeded
+* `MULTILINE_INVALID_TARGET <batch-target> <provided-target>`: mismatched PRIVMSG target within a batch
+* `MULTILINE_INVALID`: any other error
+
+## Interaction with other specifications
+
+### Message tags ([spec][message tags])
+
+Clients MUST only send non-`batch` tags on the opening BATCH command when sending multiline batches.
+
+### Message ids ([spec][message ids])
+
+Servers MUST only include a message ID on the first message of a batch when sending a fallback to non-supporting clients.
+
+### Labeled response ([draft][labeled response]) with echo-message ([spec][echo message])
+
+Servers MUST only include the label tag on the opening BATCH command when replying to a client using `echo-message`.
+
+## Client implementation considerations
+
+This section is non-normative.
+
+### Splitting long lines
+
+When splitting long lines, care is required to ensure messages are displayed appropriately when combined or in the fallback case.
+
+The recommended splitting method is to split long lines between words where possible and leave the space character at the end of the line. This ensures that concatenated lines don't lose the space, while fallback lines don't start with a space.
+
+In order to calculate the maximum allowed space before splitting a line, consider the following line:
+
+```
+:nick!~user@host PRIVMSG #channel :hello \n\r
+```
+
+The parts of a message that must fit within the 512 byte limit are as follows
+
+| part | syntax | length |
+|-|-|-|
+| leading | `:` | 1 |
+| nick | `nick` | variable |
+| nick/mask separator | `!` | 1 |
+| user | `~user` | variable |
+| user/host separator | `@` | 1 |
+| host | `host` | variable |
+| command | `PRIVMSG` | 7 |
+| channel | `#channel` | variable |
+| message separator | ` :` | 2 |
+| crlf | `\n\r` | 2 |
+| message | `hello ` | variable |
+
+Not all parts may appear in a message, but a safe limit should take them into account and include an extra safety buffer in case of any unexpected syntax added by a server, e.g. a statusmsg prefix.
+
+With a safety buffer of 10, the total message size available before a split is needed can be calculated as follows
+
+```
+512 - 14 - 10 - <nick> - <user> - <host> - <channel>
+```
+
+Clients might wish to perform a `WHO` query on themselves when connecting to a server, and keep track of numeric `396 RPL_VISIBLEHOST` to keep their `user@host` mask in sync.
+
+Clients might alternatively prefer to use a hard coded limit. A possible worst case scenario for the variable parts might be:
+
+* nick: 20
+* user: 20
+* host: 63
+* channel: 32
+
+Which would leave 353 bytes for each message.
+
+## Server implementation considerations
+
+This section is non-normative.
+
+In the context of flood protection, keep in mind that clients might send an entire batched multiline message all at once.
```

Comment:
I haven't thought this through fully, but there's an issue here with sendqs. If the sendq is small relative to the value of max-bytes, you can DoS clients by completing multiple multiline batches at once and filling up their sendqs.

Reply:
Following up on a clarifying question from #ircv3: yeah, client A can try to disconnect client B from a server by filling up its sendq. Normally, flood control is a check on this. It's not clear what the best practices are for applying flood control to multiline messages --- the way to get parity with pre-multiline behavior w.r.t. the sendq would be to apply artificial delays to the *outgoing* batch.

Reply:
If `max-lines` is not set, the worst case for a DoS on the sendq is to send `max-bytes` lines, with a single byte on each line, and `multiline-concat` on every line after the first. Assuming the longest possible nickmask, the total size of the relayed message will be about 100 times `max-bytes`. Even for a conservative value of `max-bytes`, this should be large enough to overwhelm both the sendq and the TCP send buffer (I think it's typically around 85kB on modern Linux machines).